crates/shared/src/order_validation.rs

```diff
@@ -942,16 +932,18 @@ pub fn is_order_outside_market_price(
     })
 }
 
+pub struct InvalidSigningScheme;
```

Comment:
What's the advantage of introducing this type here?
Couldn't you just as well return `ValidationError::InvalidSignature` from `convert_signing_scheme_into_quote_signing_scheme()` directly?

Reply:
As far as I know you cannot return a concrete variant from the function signature (but instead have to remove the full enum). In this case I would need to handle impossible cases (all variants that are not signatures). I could of course consider those cases `unreachable()` but I think in this case it's better to reduce the expressiveness of the function's return types (instead of requiring logic in a different module).

Reply:
But `.map_err(|_| ValidationError::InvalidSignature)` is used where the error is completely ignored. So the return type can still be `Result<QuoteSigningScheme, ValidationError>` and this `map_err(|_| ...)` can be removed, doesn't it?

Reply:
Yeah, that was my thought as well. 👍

Reply:
The problem is that this method is also used in [another place](https://github.com/cowprotocol/services/blob/b8161a5ac8530f7b35861482208bcea40f6bdb52/crates/autopilot/src/database/onchain_order_events.rs#L522) where we aren't able to do a 1:1 mapping from ValidationError (e.g. the NotFound variant no longer exists) to the required upper error type.

So if this method returned ValidationError, I would have to handle all the different error cases there and put a bunch of `unreachable()` for the error cases that cannot happen in `convert_signing_scheme_into_quote_signing_scheme`. Imo the interface of convert_signing_scheme_into_quote_signing_scheme should be minimal in if there is only one possible error type this should be it (I can also return an `Option` or `Result<_, ()>` but clippy was yelling at me for it).

Maybe I'm misunderstanding your point (as I feel we are not moving forward), in which case would you maybe be able to push a commit with your suggested change?

Reply:
> Imo the interface of convert_signing_scheme_into_quote_signing_scheme should be minimal in if there is only one possible error type this should be it

Makes sense.

> So if this method returned ValidationError, I would have to handle all the different error cases there and put a bunch of unreachable() for the error cases that cannot happen in convert_signing_scheme_into_quote_signing_scheme

But you are already completely ignoring the error that the function returns. There are no match statements [anywhere](https://github.com/cowprotocol/services/pull/3008/files#diff-1ff024ca2dfa09435b86bbecf6919e50319d070f04f71ed834341d5d254b3c15R568). With the suggested change you could at least skip one instance of `.map_err()`. But the argument that there is only 1 error condition so there should be only 1 error variant returned is good enough IMO.

Reply:
> With the suggested change you could at least skip one instance of .map_err(). 

But the other instance would become horrible, as I would need to handle cases that aren't logically possible. But in order to convince yourself they are indeed unreachable you would need to jump into another method and read the code there, which could later on change without someone remembering to update the conversion logic.